### PR TITLE
Topology-aware Environment for looking up API management endpoints and their credentials

### PR DESCRIPTION
**Pull request by heiko-koehler**
Client code should be shielded from how settings specific to an environment are injected as this mechanism might change.

The "environment" package provides abstraction for settings and all their sources to evolve independently from clients like k8s extensions or operators. Environment here should be understood as least common denominator across any consumers. Ideally the package resides in a common golang library shared between consumers like k8s extensions and operators. Most environment setting influence how clients interact with Prism, like which credentials to use. Hence prism-go-client can serve as common golang library.

Moreover multiple clients like Kubernetes clusters might share certain properties of an environment like the categories and project. However idea is that each k8s cluster environment is unique to allow infrastructure resources created on behalf of a k8s cluster to be associated with this particular k8s cluster.

We envision the following sources for environment settings:

1. "Local" using OS environment variables as source
2. "SecretDir" expecting k8s secret to be mounted as directory into container
3. "ConfigMap" expecting k8s config map to b mounted as directory into container
4. "Kubernetes" expecting access to k8s via informers for reading secrets directly

Following is the interface through which environment is accessed.

```
// Environment in which a Kubernetes has been deployed.
type Environment interface {
	// GetManagementEndpoint retrieves managment endpoint
	GetManagementEndpoint(topology Topology) (*ManagementEndpoint, error)
	// Get retrieves settings applicable to the environment like project
	// or category to which resources created on behalf of Kubernetes cluster are
	// assigned to.
	// These settings might have to be explicitly propagated to resources.
	// Return whether lookup was successful to distinguish from nil as value.
	Get(topology Topology, key string) (interface{}, error)
}
```

The same interface is implemented by all environment sources/providers.

### CODE REVIEW

File: environment/environment_test.go [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+	interface{}, error,
+) {
+	if val, ok := prov.vars[key]; ok {
+		return val, nil
+	}
+	return nil, types.ErrNotFound
+}
+
+func TestEnvironment(t *testing.T) {
+	g := NewWithT(t)
```
Comment (thunderboltsid) [inline environment/environment_test.go:34]: nit: consider using testify. If using ginkgo, please don't `.` import for readability :)

File: environment/providers/configmap/configmap.go [hunk 2]
```diff
@@ -0,0 +1,21 @@
+package configmap
```
Comment (thunderboltsid) [inline environment/providers/configmap/configmap.go:1]: Please add a `TODO` so we can keep track of the fact that configmap provider needs to be implemented.

File: environment/providers/kubernetes/types.go [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+}
+
+// NutanixPrismEndpoint defines a Nutanix API endpoint with reference to credentials.
+// Credentials are stored in Kubernetes secrets.
+type NutanixPrismEndpoint struct {
+	// address is the endpoint address (DNS name or IP address) of the Nutanix Prism Central or Element (cluster)
+	Address string `json:"address"`
+	// port is the port number to access the Nutanix Prism Central or Element (cluster)
+	Port int32 `json:"port"`
+	// use insecure connection to Prism endpoint
```
Comment (thunderboltsid) [inline environment/providers/kubernetes/types.go:72]: Consider expounding on the insecure connection i.e. this is still https but tls verification is skipped for the purposes of self signed certs.

File: environment/providers/kubernetes/kubernetes_test.go [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+	"k8s.io/client-go/informers"
+	coreinformers "k8s.io/client-go/informers/core/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/fake"
+	"k8s.io/client-go/tools/cache"
+
+	"github.com/nutanix-cloud-native/prism-go-client/environment/types"
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
```
Comment (thunderboltsid) [inline environment/providers/kubernetes/kubernetes_test.go:17]: Consider reordering to a 3-group import
```
stdlib

external

local
```

File: environment/providers/kubernetes/kubernetes.go [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+
+func (prov *provider) getCredentials(_ types.Topology) (*types.ApiCredentials, error) {
+	ref := prov.prismEndpoint.CredentialRef
+	secret, err := prov.secretInformer.Lister().Secrets(ref.Namespace).Get(ref.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	// Parse credentials in secret
+	credsData, ok := secret.Data["credentials"]
```
Comment (thunderboltsid) [inline environment/providers/kubernetes/kubernetes.go:29]: Does the secretInformer take care of base64 decoding?

File: environment/types/types.go [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+	// Address is URL of managment endpoint
+	Address *url.URL
+	// Whether to authenticate TLS endpooint
+	Insecure bool
+}
+
+// Topology is a map of topological domains to topological segments.
+// A topological domain is a sub-division of a cluster, like "region",
+// "zone", "rack", etc
+type Topology map[string]string
```
Comment (thunderboltsid) [inline environment/types/types.go:42]: Should we define Topology value types as consts? 
```go
type TopologyItem string
type Topology map[TopologyItem]TopologyItem
const (
    PrismCentralZone TopolyItem = "..."
    PrismElementCluster = "..."
)
```
or do we expect topologies to be highly flexible based on the use-case?

File: environment/providers/secretdir/secretdir_test.go [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+	g.Expect(err).To(Succeed())
+	defer func() {
+		if !t.Failed() {
+			os.RemoveAll(path)
+		}
+	}()
+	t.Logf("Temporary directory %s", path)
+	os.Setenv(envSecretDir, path)
+	g.Expect(ioutil.WriteFile(filepath.Join(path, secretKeyName),
+		[]byte("10.40.146.241:9440:admin:Nutanix.123"),
```
Comment (thunderboltsid) [inline environment/providers/secretdir/secretdir_test.go:24]: consider using random strings here :)

File: environment/environment_test.go (hunk 1 above)
Comment (heiko-koehler) [reply environment/environment_test.go:34]: I might do more than assertions :) I like the fact that Gomega can do async asserts. Also in k8s ecosystem everyone seemed to have standardized on Ginkgo + Gomega, including Kubernetes itself.

`.` import is considered idiomatic in Ginkgo and Gomega, but have now strong opinions about it.

File: environment/providers/kubernetes/kubernetes.go (hunk 5 above)
Comment (heiko-koehler) [reply environment/providers/kubernetes/kubernetes.go:29]: It does. base64 encoding is done by YAML encoder for data of type []byte. This code has been heavily influenced by CCM. See https://github.com/nutanix-cloud-native/cloud-provider-nutanix/blob/87510dd51501f288332d126a6bd9f535729edc18/pkg/provider/nutanix_client.go#L63
Idea is to consolidate common logic like parsing of Nutanix credentials across repos.

File: environment/types/types.go (hunk 6 above)
Comment (heiko-koehler) [reply environment/types/types.go:42]: That's TBD. You're right we should standardize on topology keys. For instance an availability zone might correspond to one or more PC endpoints.

File: environment/environment_test.go (hunk 1 above)
Comment (heiko-koehler) [reply environment/environment_test.go:34]: I did not make this change as code would be very verbose. Gomega (and Ginkgo for that matter) can be understood as DSLs. Functions are used like natural language. With `gomega.` prefix readability is really impacted.

File: environment/types/types.go (hunk 6 above)
Comment (thunderboltsid) [reply environment/types/types.go:44]: Cool, we can tackle that in future PRs.

File: environment/environment_test.go (hunk 1 above)
Comment (thunderboltsid) [reply environment/environment_test.go:34]: While I don't agree with the unfortunate ginkgo-everything landscape of Golang development today, I won't be gating the PR on this.

**Review by thunderboltsid — DISMISSED** [review]

File: environment/environment_test.go (hunk 1 above)
Comment (heiko-koehler) [reply environment/environment_test.go:34]: I'm falling in line with what k8s is doing. We use Ginkgo/Gomega in CNDS and k8s-e2e-test as well. I understand it's overkill for simple unit tests but simplifies writing more complex integration tests.

**Review by thunderboltsid — APPROVED** [review]

File: environment/providers/kubernetes/kubernetes_test.go [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+			Data: map[string][]byte{
+				"credentials": []byte(`
+				[
+					{
+					  "type": "basic_auth", 
+					  "data": { 
+						"prismCentral":{
+						  "username": "admin", 
+						  "password": "Nutanix.123"
+						},
```
Comment (thunderboltsid) [inline environment/providers/kubernetes/kubernetes_test.go:57]: consider using random strings here too

File: environment/providers/kubernetes/kubernetes_test.go [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+	})
+	It("must get management endpoint", func() {
+		me, err := prov.GetManagementEndpoint(nil)
+		Expect(err).To(Succeed())
+		addr, _ := url.Parse("https://10.45.43.34:9440")
+		Expect(me).To(BeEquivalentTo(&types.ManagementEndpoint{
+			ApiCredentials: types.ApiCredentials{
+				Username: "admin",
+				Password: "Nutanix.123",
+			},
```
Comment (thunderboltsid) [inline environment/providers/kubernetes/kubernetes_test.go:95]: same

File: environment/providers/local/local_test.go [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+	g.Expect(err).NotTo(Succeed())
+
+	os.Setenv(endpointEnv, "10.40.146.241:9440")
+	me, err = prov.GetManagementEndpoint(nil)
+	g.Expect(err).To(Succeed())
+	g.Expect(me.Address.String()).To(BeIdenticalTo("https://10.40.146.241:9440"))
+	g.Expect(me.ApiCredentials).To(BeIdenticalTo(types.ApiCredentials{
+		Username: "admin",
+		Password: "Nutanix.123",
+	}))
```
Comment (thunderboltsid) [inline environment/providers/local/local_test.go:38]: same

**Comment by thunderboltsid** [comment]
Please add corresponding changelog entries in https://github.com/nutanix-cloud-native/prism-go-client/blob/main/CHANGELOG.md and squash all the commits into a single commit.

**Pull request closed without merging**